### PR TITLE
Story/cite 204

### PR DESCRIPTION
**Pull request by PradnyaC11**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 

... Put ticket description here and add link to ticket ...
Implementing an 'Add Reference' Button for All Groups in Citesphere with a Search Field Feature to Easily Add References to Group's Reference Field

https://diging.atlassian.net/browse/CITE-204

## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Autowired
+    private SearchEngine engine;
+
+    @RequestMapping(value = { "/auth/group/{zoteroGroupId}/find/references" })
+    public ResponseEntity<ResultPage> getReferences(Authentication authentication, @PathVariable String zoteroGroupId,
+            @RequestParam(value = "searchTerm", required = false) String searchTerm,
+            @RequestParam(defaultValue = "0", required = false, value = "page") int page,
+            @RequestParam(defaultValue = "20", required = false, value = "pageSize") int pageSize) {
+        return new ResponseEntity<ResultPage>(engine.search(searchTerm, zoteroGroupId, Math.max(page - 1, 0), 50), HttpStatus.OK);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java:32]: what's 50? The pagesize? that should not be hard-coded

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import edu.asu.diging.citesphere.core.exceptions.CitationIsOutdatedException;
+import edu.asu.diging.citesphere.core.exceptions.GroupDoesNotExistException;
+import edu.asu.diging.citesphere.core.exceptions.ZoteroHttpStatusException;
+import edu.asu.diging.citesphere.core.exceptions.ZoteroItemCreationFailedException;
+import edu.asu.diging.citesphere.core.service.ICitationManager;
+import edu.asu.diging.citesphere.model.bib.ICitation;
+import edu.asu.diging.citesphere.user.IUser;
+
+@Controller
+public class UpdateItemController {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:27]: We have an `EditItemController`, this name is just confusing. It's a very specific update it does, isn't it? So the name should reflect that.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private ICitationManager citationManager;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/update", method = RequestMethod.POST)
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:34]: same thing, this url is too confusing.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public ResponseEntity<?> updateReference(Authentication authentication,
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:35]: bad method name: the method does not update the reference

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            @PathVariable("zoteroGroupId") String zoteroGroupId, @PathVariable("itemId") String itemId,
+            @RequestParam(value = "referenceCitationKey") String referenceCitationKey,
+            @RequestParam(value = "reference") String reference) {
+        try {
+            ICitation citation = citationManager.getCitation((IUser) authentication.getPrincipal(), zoteroGroupId,
+                    itemId);
+            citation = citationManager.updateCitationReference(citation, referenceCitationKey, reference);
+            citationManager.updateCitation((IUser) authentication.getPrincipal(), zoteroGroupId, citation);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:43]: what do `updateCitationReference` and `updateCitation` do? why do they both need to be called? if they do things that should always be done together, then this is business logic and should not be in the controller.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            return new ResponseEntity<>("{\"error\": \"Item " + itemId + " not found.\"}", HttpStatus.NOT_FOUND);
+        } catch (ZoteroHttpStatusException e) {
+            logger.error("Zotero threw exception.", e);
+            return new ResponseEntity<>("{\"error\": \"" + e.getMessage() + "\"}", HttpStatus.INTERNAL_SERVER_ERROR);
+        } catch (ZoteroConnectionException e) {
+            logger.error("Zotero connection failed.", e);
+            return new ResponseEntity<>("{\"error\": \"" + e.getMessage() + "\"}", HttpStatus.INTERNAL_SERVER_ERROR);
+        } catch (CitationIsOutdatedException e) {
+            logger.error("Citation is outdated.", e);
+            return new ResponseEntity<>("{\"error\": \"Item " + itemId + " is outdated.\"}", HttpStatus.NOT_FOUND);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:60]: this should probably be a 409 Conflict

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 7]
```diff
@@ -1,31 +1,38 @@
...
 			var clicked = e.target;
-			$("#ref-unparsed").text($(clicked).text());
-			$("#parsed-ref-authors").text($(clicked).attr('data-authors'));
-			$("#parsed-ref-title").text($(clicked).attr('data-title'));
-			$("#parsed-ref-year").text($(clicked).attr('data-year'));
-			var identifier = $(clicked).attr('data-identifier');
-			var identType = $(clicked).attr('data-identifier-type');
-			if (identType != undefined && identType != "") {
-				identifier = identifier + " (" + identType + ")";
+			citationID = $(clicked).attr('data-citation-id');
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:9]: this creates a global variable, doesn't it? That should not be needed.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 8]
```diff
@@ -1,31 +1,38 @@
...
+			if(citationID != undefined && citationID != "") {
+				zoteroGroupId = $(clicked).attr('data-zotero-group-id') 
+				window.location.href = "../items/" + citationID;
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:12]: no, needs to be a properly generated url

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java (hunk 5 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:43]: updateCitation method is used at multiple places so, didn't change it. The method saves changed citation data to zetero.  updateCitationReference name is changed to addCitationToReferences, which adds found citation to references of the editing citation.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java (hunk 2 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:27]: Changed name of class to AddReference Controller

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java (hunk 3 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:34]: changed the url from update to addReference

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java (hunk 4 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:35]: Changed updateReference to addReference

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 9]
```diff
@@ -489,9 +501,25 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
     }
-    
+
     @Override
     public void deleteLocalGroupCitations(String groupId) {
         citationStore.deleteCitationByGroupId(groupId);
     }
+
+    @Override
+    public ICitation updateCitationReference(ICitation citation, String referenceCitationKey, String reference) {
```
[the remark this answers is not in the document]
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:511]: Name of method changed to addCitationToReferences

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 10]
```diff
@@ -489,9 +501,25 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
+        }
+        if (!citation.getKey().equals(referenceCitationKey)) {
```
[the remark this answers is not in the document]
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:516]: It checks if the citation is added to its own reference list.  Added an exception for this condition.

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 11]
```diff
@@ -494,4 +504,22 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
     }
+
+    @Override
+    public ICitation addCitationToReferences(ICitation citation, String referenceCitationKey, String reference) throws SelfCitationException {
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
+        }
+        if (!citation.getKey().equals(referenceCitationKey)) {
+            IReference iReference = new Reference();
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:515]: why is this called `iReference`?

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 12]
```diff
@@ -84,6 +90,9 @@ public class CitationManager implements ICitationManager {
     @Autowired
     private IAsyncCitationProcessor asyncCitationProcessor;
 
+    @Autowired
+    private CitationRepository citationRepository;
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:94]: this should not be used here directly but the `citationStore`

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/addReference", method = RequestMethod.POST)
+    public ResponseEntity<?> addReference(Authentication authentication,
+            @PathVariable("zoteroGroupId") String zoteroGroupId, @PathVariable("itemId") String itemId,
+            @RequestParam(value = "referenceCitationKey") String referenceCitationKey,
+            @RequestParam(value = "reference") String reference) {
+        try {
+            ICitation citation = citationManager.getCitation((IUser) authentication.getPrincipal(), zoteroGroupId,
+                    itemId);
+            citation = citationManager.addCitationToReferences(citation, referenceCitationKey, reference);
+            citationManager.updateCitation((IUser) authentication.getPrincipal(), zoteroGroupId, citation);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:44]: why does the citation needs to be updated? has anything changed?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java (hunk 11 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:515]: IReference is interface and it's implementation is Reference for reference of the citation. It was already present in citesphere-model.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java (hunk 13 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:44]: updateCitation method is called here to update the newly added reference in zetero too. Removing this does not add the reference to zetero's notes.

Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:44]: `addCitationToReferences` should save the citation, there shouldn't be an extra call needed (especially not form the controller).

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:1]: please read and follow our testing guidelines/naming conventions

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+
+import edu.asu.diging.citesphere.core.search.service.SearchEngine;
+import edu.asu.diging.citesphere.core.search.service.impl.ResultPage;
+
+@Controller
+@PropertySource("classpath:/config.properties")
+public class ReferenceController {
```
[the remark this answers is not in the document]
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java:20]: Changed name of controller to SearchReferencesController, method name to - searchReferences  and URL to - /auth/group/{zoteroGroupId}/references?searchTerm={searchTerm}

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 15]
```diff
@@ -494,4 +500,22 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
     public void deleteLocalGroupCitations(String groupId) {
         citationStore.deleteCitationByGroupId(groupId);
     }
+
+    @Override
+    public ICitation addCitationToReferences(ICitation citation, String referenceCitationKey, String reference) throws SelfCitationException {
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: if this is the case, then `references` will never be added to the citation.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java [hunk 16]
```diff
@@ -38,6 +38,7 @@ public String getItem(Authentication authentication, Model model, @PathVariable(
             @RequestParam(required = false, defaultValue = "", value = "conceptIds") String[] conceptIds) throws GroupDoesNotExistException, CannotFindCitationException, ZoteroHttpStatusException {
         ICitation citation = citationManager.getCitation((IUser)authentication.getPrincipal(), zoteroGroupId, itemId);
         model.addAttribute("zoteroGroupId", zoteroGroupId);
+        model.addAttribute("itemId", itemId);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:41]: why is this needed? `itemId` is the id of the citation, isn't it? And the citation object is already in the model.

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java [hunk 17]
```diff
@@ -69,7 +76,13 @@ public class CitationManagerTest {
     ICitationGroup group;
     private ICitationGroup group1;
     private ICitationGroup group2;
-        
+    private ICitation citation;
+    private final String REFERENCE = "reference";
+    private final String CITATION_KEY = "citationKey";
+    private ICitation updatedCitation;
+    private  Set<IReference> references;
+    private IReference iReference;
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:84]: `iReference` is a bad variable name. What would the `i` stand for?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
- layout is off
image

- The author for the references string seems to be taken randomly
image
The next time it's "Haines R, 2019, What Makes Research Software Sustainable? An Interview Study With Research Software Engineers"
If there are multiple authors, it should be "First author et al."
- then clicking "Add", it takes a long time where the browser is not reacting. If the user clicks again, it adds the reference twice. There needs to be a loading indicator and the user should not be able to click the button again.
- And I can add the same item as reference

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java (hunk 15 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: This condition is added for the citations that don't have any references already added to them. In that case, there's a probability the references might be set to null. It occurs when a person create the citation, instead of importing it.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java (hunk 16 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:41]: Removed itemId.

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java (hunk 15 above)
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: yes, I understand that, but right now the references of the citation object will still be null (you never set the references on the citation, just the local variable)

Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: Removing the condition gave me null pointer exception error when I tried to add reference to a dummy citation that I created in zetero. It does not give error when the new dummy citation is created in citesphere, because then it's initialized to empty array.

**Comment by PradnyaC11** [comment]
> * layout is off
>   image
> * The author for the references string seems to be taken randomly
>   image
>   The next time it's "Haines R, 2019, What Makes Research Software Sustainable? An Interview Study With Research Software Engineers"
>   If there are multiple authors, it should be "First author et al."
> * then clicking "Add", it takes a long time where the browser is not reacting. If the user clicks again, it adds the reference twice. There needs to be a loading indicator and the user should not be able to click the button again.
> * And I can add the same item as reference

1. Updated the layout. 
2. For random author creation problem, I have sort the list by firstname. 
3. Added the loaded spinner. 
4. I tested adding same item as reference, It threw exception as expected. I tested it on item - "Self-Organization in Pattern Formation".

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java (hunk 15 above)
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: What I'm saying is that this code:
```
Set<IReference> references = citation.getReferences();
if (references == null) {
            references = new HashSet<>();
}
...
 references.add(newReference);
```
will create a list of references that you can add to. However, if `citation.getReferences()` returned null before the if, it will still return null after the if. The following code wont' throw a nullpointer, but the changes you make to `references` won't be stored.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java (hunk 13 above)
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:44]: still still

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 18]
```diff
@@ -104,7 +115,170 @@
...
+						}
+						if (firstName) {
+							reference += firstName;
+						}
+						if(authors.length > 1) {
+							reference += " et al."
+						}
+						reference += ", "
+					}
+	                reference += (citationItem["dateFreetext"] != null && citationItem["dateFreetext"] != "") ? citationItem["dateFreetext"] + ", " : "";
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:204]: all of this should not be done in javascript. reuse the code that generates the titles on the item pages here.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jdamerow) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java]: But it searches citations not references.

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html (hunk 18 above)
Comment (PradnyaC11) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:204]: Created a template in the same file, as short form of citation is not being used at any other place and used it for title and at this place.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java (hunk 13 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:44]: Moved it to addCitationReference.

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java (hunk 15 above)
Comment (PradnyaC11) [reply citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: Got the logical error here. I have added a line to set references of citation so that the case you are saying won't happen.

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 19]
```diff
@@ -494,4 +498,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
+        }
+        if (!citation.getKey().equals(referenceCitationKey)) {
+            IReference newReference = new Reference();
+            newReference.setCitationId(referenceCitationKey);
+            newReference.setReferenceString(reference);
+            references.add(newReference);
+            citation.setReferences(references);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:516]: this should happen in line 510

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+@Controller
+public class AddReferenceController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private ICitationManager citationManager;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/addReference", method = RequestMethod.POST)
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:34]: should be `.../references/add`

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/SearchCitationsController.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+@Controller
+@PropertySource("classpath:/config.properties")
+public class SearchCitationsController {
+
+    @Autowired
+    private SearchEngine engine;
+
+    @RequestMapping(value = { "/auth/group/{zoteroGroupId}/references" })
+    public ResponseEntity<ResultPage> searchReferences(Authentication authentication, @PathVariable String zoteroGroupId,
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/SearchCitationsController.java:24]: similar to the class name, this should not be called 'searchReferences' since it's not searching references, and the url should similarly be reflecting that it searches citations. Right now the url suggests it returns references.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
compilation errors in tests

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 22]
```diff
@@ -104,7 +150,143 @@
...
+	            x: 50,
+	            y: 90
+	        },
+	        timer: time,
+	        animate: {
+	            enter: 'animated fadeInRight',
+	            exit: 'animated fadeOutRight'
+	        },
+	    });
+	}
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:289]: there are a couple of odd indentation issues

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- The search results box that shows the reference search results needs to show author and year, not just title.
- Also the box needs to say "Reference Search Result**s**"
- And the column with the add button should have a heading "Add as reference"
- The references input field in the "Add references" modal should be deleted; instead the modal should close when an add button is clicked.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
When clicking on a reference after it has been added, the pop up is all empty. it needs to pull the information from the citation.

**Comment by jdamerow** [comment]
and there is a blank that shouldn't be there:
image

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 23]
```diff
@@ -494,4 +496,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
+        if (references == null) {
+            references = new HashSet<>();
+            citation.setReferences(references);
+        }
+        if (!citation.getKey().equals(referenceCitationKey)) {
+            IReference newReference = new Reference();
+            newReference.setCitationId(referenceCitationKey);
+            newReference.setReferenceString(reference);
+            references.add(newReference);
+        } else {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:515]: let's turn this if/else around so that we loose one indented block, so
```
if keys are equal:
      throw exception
code if they are not equal goes here
```

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 24]
```diff
@@ -494,4 +496,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
+    public ICitation addCitationToReferences(IUser user, ICitation citation, String zoteroGroupId, 
+            String referenceCitationKey, String reference) throws SelfCitationException, 
+            ZoteroConnectionException, CitationIsOutdatedException, ZoteroHttpStatusException, 
+            ZoteroItemCreationFailedException {
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
+            citation.setReferences(references);
+        }
+        if (citation.getKey().equals(referenceCitationKey)) {
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:510]: this should happen first thing. Nothing of the code above needs to be executed if an exception is thrown here.

File: citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java [hunk 25]
```diff
@@ -494,4 +496,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
...
+
+    @Override
+    public ICitation addCitationToReferences(IUser user, ICitation citation, String zoteroGroupId, 
+            String referenceCitationKey, String reference) throws SelfCitationException, 
+            ZoteroConnectionException, CitationIsOutdatedException, ZoteroHttpStatusException, 
+            ZoteroItemCreationFailedException {
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
+            citation.setReferences(references);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:508]: merge with line 507

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 26]
```diff
@@ -104,7 +134,141 @@
...
+	            var message = "Reference added successfully";
+	            showNotification(message, "success", 1000);
+	        },
+	        error: function(data) {
+	            showNotification(" Error while updating reference!", "danger");
+	        }, 
+			complete: function() {
+				$("#addReferenceSpinner").hide();
+				$("#referenceModal").modal('hide');
+			}
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:257]: indentation

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 27]
```diff
@@ -104,7 +134,141 @@
...
+	    
+	    $("#closeReferenceSearchResult").click(function() {
+	        $("#selectReferenceModal").modal('hide');
+	        $("#searchReferenceSpinner").hide();
+	        $("#addReferenceSpinner").hide();
+	    });
+	    
+	    $("#addReferences").click(function() {
+	        $("#citationReference").val('');
+			$("#itemReference")[0].innerHTML = '';
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:172]: indentation

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 28]
```diff
@@ -104,7 +134,141 @@
...
+	    
+	    $("#closeReferenceModalBtn").click(function() {
+	    	$("#referenceModal").modal('hide');
+	    });	    
+
+	    $("#addReferenceButton").click(function() {
+	        $("#addReferenceSpinner").show();
+	    	document.getElementById('addReferenceSpinner').style.display = 'inline-block';
+
+			var zoteroGroupId = $("#addReferences").data('zotero-group-id');
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:146]: indent

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html (hunk 22 above)
Comment (jdamerow) [reply citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:271]: I pointed out a few but there are a few more

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- When there are no search results, there should be a message that says so.
- search should also happen if you hit enter after entring a term

**Comment by jdamerow** [comment]
Make it so, Jenkins.

File: citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html [hunk 29]
```diff
@@ -104,7 +134,156 @@
...
+	        type: 'GET',
+	        url: url,
+	        async: false,
+	        success: function(data) {
+	            var results = data["results"];
+	            if(results.length==0) {
+	            	$("#citationSearchTable").empty();
+	            	var content = '<div class="alert warning">'
+	            	content += '<strong>Warning!</strong>'
+	            	content += ' No citations found for the search query. Try other search query.'
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:197]: The message should just be: "No citations were found for the entered search term."
(No "warning" no "try other search query")

**Review by jdamerow — CHANGES_REQUESTED** [review]
After that should be good.

**Comment by jdamerow** [comment]
Also there is a bug, if I search for soemthing that yields results, it shows results. I click close without selecting something. I search for something that doesn't exist, it shows me no results. I click close. I search for something that should yield results, but now it shows me no results.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Another bug:
- I search for a term, it shows results
- I select one
- I don't click "Add reference" but search again
- I don't select anything but just click "close"
- the reference I selected before is still shown, but the 'add reference' button is now disabled
=> I should be able to select the reference I selected first
